### ISSUE TITLE
ZCA­007 - Forever growing nullifier set will end up being stored in  nonvolatile memory 

### ISSUE THREAD

**Issue by nathan-at-least** [zcash/zcash#1390]

**Comment by nathan-at-least** [comment]
This is the primary known protocol scaling issue in our design. Fixing it is an open research question. We believe this won't be a concern for at least 2 years.

**Comment by daira** [comment]
I honestly don't see how this is even a problem, in principle. Each nullifier is 32 bytes. Suppose that we have 2 MB blocks full of JoinSplits for 10 years. A JoinSplit is 1802 bytes, so that is at most 1163 JoinSplits per 2.5 minutes. That makes at most 1163 × (60/2.5) × 24 × 365.25 × 10 × 2 × 32 bytes, which is about 156.6 GB, as the maximum size of the nullifier set after 10 years. Of course, this makes a number of assumptions: e.g. that we don't reduce the size of JoinSplits, or increase the number of JoinSplit outputs or the maximum transaction size. I think it will be relatively straightforward and cheap to store 156 GB 10 years from now. Indeed, under these (very conservative!) assumptions, I think it is extremely likely that the trend in storage capacity and price will mean that storing the nullifier set on secondary storage is *never* a serious problem within the life of Zcash.

(The commitment tree would run out first, but for the sake of argument let's assume we expand that as necessary.)

**Comment by SergioDemianLerner** [comment]
I agree the issue has low priority.  But if the nullifyiers set becomes higher than the available RAM (e.g. 4 GB), verifying transactions will require at least two random disk accesses (or about 30 msec on a non-SDD drive ?). Then transaction execution time will not depend solely on CPU speed.

**Comment by daira** [comment]
Non-SSD drives will be obsolete then.
